src/prototype_creep_startup_tasks.js

```diff
@@ -133,6 +133,35 @@ Creep.repairStructure = function(creep) {
   return creep.repairStructure();
 };
 
+Creep.prototype.getEnergyFromAnyStructures = function() {
+  if (this.memory.role !== 'builder' && this.memory.role !== 'universal') {
+    return false;
+  }
+  if (!this.room.isConstructingSpawn()) {
+    return false;
+  }
+  let structures = this.room.findPropertyFilter(FIND_STRUCTURES, 'structureType', [STRUCTURE_CONTROLLER, STRUCTURE_RAMPART, STRUCTURE_EXTRACTOR, STRUCTURE_OBSERVER], {
+    inverse: true,
+    filter: Room.structureHasEnergy,
+  });
+  if (this.carry.energy || !structures.length) {
+    return false;
+  }
+  // Get energy from the structure with highest amount first
```

Comment:
why no the lowest?

Reply:
Because that means 2 things I want to avoid.
1. it would use Storage at last, and Storage have quite some energy and should be extracted out then use for the hard time.
2. lowest sometimes not worthy to walt to and extract, they just have too less energy. when a spawn is destroyed, I think we must rebuild it as fast as possible, so wasting time on the road seems not quite wise to me.

Reply:
i don't know if i like that... getEnergyFromAnyStructures means even hostile storage
i sometimes find a left room let's say hostile, with an inactive player, so i claim the room.
what i don't want is that the nextroomers grab the energy from it. 

what i like more is that the universal grab the energy and build the extensions, towers and upgrade controller and when they reach the RoomControllerLevel 4 they need to rebuild the storage anyway

Reply:
> i don't know if i like that... getEnergyFromAnyStructures means even hostile storage i sometimes find a left room let's say hostile, with an inactive player, so i claim the room. what i don't want is that the nextroomers grab the energy from it.
> 
> what i like more is that the universal grab the energy and build the extensions, towers and upgrade controller and when they reach the RoomControllerLevel 4 they need to rebuild the storage anyway

Come to think of it, It is actually OK for me to accept only grab energy from own structure, as it is hard to see any hostile structure in my waiting spawn room.

Will change it to only grab energy from mystructure.